src/zkevm_specs/ecc_circuit.py

```diff
@@ -0,0 +1,261 @@
+        raise NotImplementedError("assign_mul is not supported yet")
+
+    @classmethod
+    def assign_pairing(cls, p: List[Tuple[Word, Word]], out: Tuple[Word, Word]):
+        raise NotImplementedError("assign_pairing is not supported yet")
+
+    def verify(
+        self, cs: ConstraintSystem, max_add_ops: int, max_mul_ops: int, max_pairing_ops: int
+    ):
+        # Copy constraints between EccTable and ECCVerifyChip
+        cs.constrain_equal_word(Word(self.ecc_chip.p0[0].n), self.row.px)
+        cs.constrain_equal_word(Word(self.ecc_chip.p0[1].n), self.row.py)
+        cs.constrain_equal_word(Word(self.ecc_chip.p1[0].n), self.row.qx)
+        cs.constrain_equal_word(Word(self.ecc_chip.p1[1].n), self.row.qy)
+        cs.constrain_equal(self.ecc_chip.output[0], self.row.out_x)
+        cs.constrain_equal(self.ecc_chip.output[1], self.row.out_y)
+
+        is_add = cs.is_equal(self.row.op_type, FQ(EccOpTag.Add))
+        is_mul = cs.is_equal(self.row.op_type, FQ(EccOpTag.Mul))
+        is_pairing = cs.is_equal(self.row.op_type, FQ(EccOpTag.Pairing))
+        # Must be one of above operations
+        cs.constrain_equal(is_add + is_mul + is_pairing, FQ(1))
+
+        num_add = 0
+        num_mul = 0
+        num_pairing = 0
+        if is_add == FQ(1):
+            num_add += 1
+            assert (
+                num_add <= max_add_ops
+            ), f"exceeds max number of add operation, max_add_ops: {max_add_ops}"
+            self.verify_add(cs)
+
+        if is_mul == FQ(1):
+            num_mul += 1
+            assert (
+                num_mul <= max_mul_ops
+            ), f"exceeds max number of mul operation, max_mul_ops: {max_mul_ops}"
+            self.verify_mul(cs)
+
+        if is_pairing == FQ(1):
+            num_pairing += 1
+            assert (
+                num_pairing <= max_pairing_ops
+            ), f"exceeds max number of pairing operation, max_pairing_ops: {max_pairing_ops}"
+            self.verify_pairing(cs)
+
+    def verify_add(self, cs: ConstraintSystem):
+        # input_rlc is zero bcs it's only used in pairing
+        cs.constrain_zero(self.row.input_rlc)
+
+        cs.constrain_equal(FQ(self.ecc_chip.verify_add()), self.row.is_valid)
+
+    def verify_mul(self, cs: ConstraintSystem):
+        # input_rlc is zero bcs it's only used in pairing
+        cs.constrain_zero(self.row.input_rlc)
+        # qy is zero bcs q is scalar in ecMul so we only use qx
+        cs.constrain_zero(self.row.qy)
+
+        cs.constrain_equal(FQ(self.ecc_chip.verify_mul()), self.row.is_valid)
+
+    def verify_pairing(self, cs: ConstraintSystem):
+        # p and q are all zero. All input points are RLCed and stored in input_rlc
+        cs.constrain_zero(self.row.px)
+        cs.constrain_zero(self.row.py)
+        cs.constrain_zero(self.row.qx)
+        cs.constrain_zero(self.row.qy)
+        # output of pairing is either 0 or 1 and it stores in low part
```

Comment:
Nitpick: "and stored in the lower part"?

Reply:
fixed in [bece499](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/bece499b09db05ca3ce5b2339a24a6ca37690427)